File: src/main/java/com/cinebook/service/NotificationService.java
package com.cinebook.service;

import com.cinebook.dto.response.NotificationResponse;
import com.cinebook.dto.response.PageResponse;
import com.cinebook.dto.response.UnreadCountResponse;
import com.cinebook.entity.Booking;
import com.cinebook.entity.User;
import com.cinebook.enums.NotificationType;
import org.springframework.data.domain.Pageable;

public interface NotificationService {

    /**
     * Lấy danh sách thông báo phân trang của người dùng hiện tại (mới nhất trước).
     */
    PageResponse<NotificationResponse> getUserNotifications(String userId, Pageable pageable);

    /**
     * Lấy số lượng thông báo chưa đọc của người dùng hiện tại.
     */
    UnreadCountResponse getUnreadCount(String userId);

    /**
     * Đánh dấu một thông báo cụ thể là đã đọc (bảo đảm quyền sở hữu).
     */
    NotificationResponse markAsRead(String notificationId, String userId);

    /**
     * Đánh dấu toàn bộ thông báo của người dùng là đã đọc.
     */
    void markAllAsRead(String userId);

    /**
     * Tạo thông báo mới cho sự kiện nghiệp vụ (Idempotent với bảo vệ Unique Constraint).
     */
    void createNotification(User user, Booking booking, NotificationType type, String title, String message);

    /**
     * Tạo thông báo mới bằng User ID và Booking ID.
     */
    void createNotification(String userId, String bookingId, NotificationType type, String title, String message);
}

File: frontend/src/services/notification.service.ts
import apiClient from './api'
import type { PageResponse } from '@/types/api.types'
import type {
  NotificationResponse,
  UnreadCountResponse,
} from '@/types/notification.types'

export const notificationService = {
  /**
   * Lấy danh sách thông báo phân trang của người dùng hiện tại
   */
  async getNotifications(page = 0, size = 10): Promise<PageResponse<NotificationResponse>> {
    const response = await apiClient.get<PageResponse<NotificationResponse>>('/api/v1/notifications', {
      params: { page, size },
    })
    return response.data
  },

  /**
   * Lấy số lượng thông báo chưa đọc của người dùng hiện tại
   */
  async getUnreadCount(): Promise<UnreadCountResponse> {
    const response = await apiClient.get<UnreadCountResponse>('/api/v1/notifications/unread-count')
    return response.data
  },

  /**
   * Đánh dấu một thông báo là đã đọc
   */
  async markAsRead(id: string): Promise<NotificationResponse> {
    const response = await apiClient.patch<NotificationResponse>(`/api/v1/notifications/${id}/read`)
    return response.data
  },

  /**
   * Đánh dấu tất cả thông báo của người dùng là đã đọc
   */
  async markAllAsRead(): Promise<void> {
    await apiClient.patch<void>('/api/v1/notifications/read-all')
  },
}

